### ISSUE TITLE
[Refactor] Abstract GnoVM from the System

### ISSUE THREAD

**Issue by ilgooz** [gnolang/gno#734]
## Description

The packages `gnovm/gnolang` and `gnovm/tests` currently have dependencies on system I/O and platform-specific code, which makes it difficult to compile these packages to WASM and run them in the browser to eventually execute unit and filetests on the browser for realms and packages, and provide native debugger and state inspection features to GNO IDE. The goal is to make these packages WASM-compatible and simplify the code by introducing an appropriate level of abstraction, abstracting away from the underlying system and making the packages contain only pure logic.

Specifically, the following issues have been identified:

1. The packages depend on a LevelDB package that has platform-specific code and does not support the `goarch=wasm goos=js` target.
2. The packages use the system's filesystem API directly.
3. The packages execute several commands, such as `go list`, directly.
4. The packages depend on `net/http/pprof`.

## Suggestions

To make the corresponding packages more abstract and WASM-compatible, the following suggestions are proposed:

1. Instead of making direct system calls, depend on the `io/fs` interface.
2. Create a minimal interface (`Database`) to abstract away the LevelDB implementation.
3. Create minimal interfaces to abstract away command executions.
4. Remove the direct dependence on `net/http/pprof`.
5. Consider putting the actual implementations in separate package(s) and providing them in the caller package (`gnovm/cmd/gno`).

By implementing these suggestions, we can make the packages more modular, easier to test and WASM-compatible.

**Comment by moul** [comment]
Thanks for this suggestion. I agree that creating this abstraction is essential for better compatibility of GnoVM across various environments, including wasm. Let's go for it.

> The packages execute several commands, such as `go list`, directly.

Isn't this limited to `gnovm/cmd/gnodev` only?

> Create a minimal interface (Database) to abstract away the LevelDB implementation.

Yes, but be cautious of excessive minimalism here. Minimal implementation may hinder transaction rollbacks. Consider having multiple calls with transaction logic instead of moving it to the top-level.

> Consider putting the actual implementations in separate package(s) and providing them in the caller package (gnovm/cmd/gno).

Need more details before proceeding, please.

**Comment by ilgooz** [comment]
> Isn't this limited to gnovm/cmd/gnodev only?

I found one instance at [gnovm/pkg/gnolang](https://github.com/gnolang/gno/blob/5b275ff5ff70aa4fab52b03893182ec337d7a751/gnovm/pkg/gnolang/precompile.go#L83) package, as `gnodev` is one the callers of this package, it is probably not the only one.

> Yes, but be cautious of excessive minimalism here. Minimal implementation may hinder transaction rollbacks. Consider having multiple calls with transaction logic instead of moving it to the top-level.

I totally endorse this and the interface signature should be discussed before starting to an implementation. 

> Need more details before proceeding, please.

Once we make the subject packages depend on interfaces, we'll need to provide a separate place to host the actual implementations and then provide them to the dependent packages. 

Here's an example of how we could approach this:

1. Modify the `gnovm/pkg/gnolang` package to depend on the `Database` interface instead of directly importing the LevelDB package. Define the `Database` interface and modify `gnolang.NewMachine` to accept an implementation of `Database` as an argument.
4. Ensure that `gnovm/pkg/gnolang` is not aware of LevelDB, but depends only on the `gnovm/pkg/gnolang.Database` interface.
5. Create a new isolated package called `gnovm/pkg/gnolang/db`, which implements `Database` by using the LevelDB dependency.
6. In `gnovm/cmd/gno`, initialize the LevelDB implementation of `Database` by importing `gnovm/pkg/gnolang/db` and provide it to `gnovm/pkg/gnolang.NewMachine`. Apply the same to other places that initializes a `Machine`.
7. Then in the WASM-compiled version of `Machine` type, we can provide a JavaScript implementation of `Database` to the `NewMachine` function.

In addition, we could create small PRs to address different points in the Description. For example, we could create a separate PR to modify `NewMachine` to depend on a `io/fs` virtual file system and provide an implementation for it that actually interact with the OS Filesystem API.

**Comment by moul** [comment]
> I found one instance at [gnovm/pkg/gnolang](https://github.com/gnolang/gno/blob/5b275ff5ff70aa4fab52b03893182ec337d7a751/gnovm/pkg/gnolang/precompile.go#L83) package, as gnodev is one the callers of this package, it is probably not the only one.

Yes, good catch, btw it's easy to replace and could eventually be removed entirely thanks to `gnomod`.

> I totally endorse this and the interface signature should be discussed before starting to an implementation.

Great, count on me for the review 👍 

> Once we make the subject packages depend on interfaces, we'll need to provide a separate place to host the actual implementations and then provide them to the dependent packages.
> ...

Ok, I see; please ping me when you have a proposal for the final structure, thank you.

**Comment by ilgooz** [comment]
This issue isn't relevant anymore.

We went with implementing OS level file system bindings which won't require any change in code like introducing a virtual file system. 

Once the following issues / PRs are solved, gnovm should have a full support for `js+wasm` target:
* https://github.com/gnolang/gno/issues/1008
* https://github.com/gnolang/gno/pull/1006
